pyvista/utilities/transformations.py

```diff
@@ -1,20 +1,214 @@
+    r"""Return a 4x4 matrix for rotation about any axis by given angle.
+
+    Rotations around an axis that contains the origin can easily be
+    computed using Rodrigues' rotation formula. The key quantity is
+    the ``K`` cross product matrix for the unit vector ``n`` defining
+    the axis of the rotation:
+
+             /   0  -nz   ny \
+        K =  |  nz    0  -nx |
+             \ -ny   nx    0 /
+
+    For a rotation angle ``phi`` around the vector ``n`` the rotation
+    matrix is given by
+
+        R = I + sin(phi) K  + (1 - cos(phi)) K^2
+
+    where ``I`` is the 3-by-3 unit matrix and ``K^2`` denotes the matrix
+    square of ``K``.
+
+    If the rotation axis doesn't contain the origin, we have to first
+    shift real space to transform the axis' ``p0`` reference point into
+    the origin, then shift the points back after rotation:
+
+        p' = R @ (p - p0) + p0 = R @ p + (p0 - R @ p0)
+
+    This means that the rotation in general consists of a 3-by-3
+    rotation matrix ``R``, and a translation given by
+    ``b = p0 - R @ p0``. These can be encoded in a 4-by-4 transformation
+    matrix by filling the 3-by-3 leading principal submatrix with ``R``,
+    and filling the top 3 values in the last column with ``b``.
+
+    Parameters
+    ----------
+    axis : 3-length sequence
+        The direction vector of the rotation axis. It need not be a
+        unit vector, but it must not be a zero vector.
+
+    angle : float
+        Angle of rotation around the axis. The angle is defined as a
+        counterclockwise rotation when facing the normal vector of the
+        rotation axis. Passed either in degrees or radians depending on
+        the value of ``deg``.
+
+    point : 3-length sequence, optional
+        The origin of the rotation (a reference point through which the
+        rotation axis passes). By default the rotation axis contains the
+        origin.
+
+    deg : bool, optional
+        Whether the angle is specified in degrees. ``False`` implies
+        radians.
+
+    Examples
+    --------
+    Generate a transformation matrix for rotation around a cube's body
+    diagonal by 120 degrees.
+
+    >>> import numpy as np
+    >>> from pyvista import transformations
+    >>> trans = transformations.axis_angle_rotation([1, 1, 1], 120)
+
+    Check that the transformation cycles the cube's three corners.
+
+    >>> corners = np.array([
+    ...     [1, 0, 0],
+    ...     [0, 1, 0],
+    ...     [0, 0, 1],
+    ... ])
+    >>> rotated = transformations.apply_transformation_to_points(trans, corners)
+    >>> np.allclose(rotated, corners[[1, 2, 0], :])
+    True
+
+    """
     if deg:
+        # convert to radians
         angle *= np.pi / 180
-    import transforms3d as tf3d
-    return tf3d.axangles.axangle2aff(axis, angle, point=point)
+
+    # return early for no rotation
+    if np.isclose(angle % (2 * np.pi), [0, 2 * np.pi], rtol=0, atol=1e-10).any():
```

Comment:
Good check for this.

Reply:
Is this desirable? I spend a lot of time in optimization world where gradients are preferably smooth. This if statement introduces a flat  "notch" in the gradient for `angle` = +/- 1e-10. A quick test shows that there is no actual singularity in the transform at `angle = 0`

Reply:
@whophil, I didn't consider this.  @adeak, we may consider dropping this.

Reply:
@whophil what do you mean by "gradients"? This is just an attempt to check that the rotation angle is `n * 2 * pi`, in which case we don't have to do anything. For all intents and purposes a rotation with 1e-10 degrees is not a rotation at all, especially since VTK uses single precision for points if I recall correctly.

This is purely a performance optimization, so we can just remove it or be more strict with `atol`. I just don't (yet) understand the situation where this breaks something.

Reply:
Sorry bad explanation. All I mean is that I would expect all elements of the transform matrix (and point transformations using it) to be smooth w.r.t. angle. With the if statement, I don't think that will be true.

Regarding precision, the output of this function could conceivably be used on non-VTK (and therefore double prec) points, so an assumption about precision seems out of place.

Reply:
I see, thanks for clarifying.

Indeed, the true rotation is smooth, and with the if this is not technically the case. But we're working with technology where
```py
>>> 0.1 + 0.2 == 0.3
False
```

And where 21 turns are no longer a complete turn
```py
>>> (21 * 2*np.pi) % (2 * np.pi)
7.105427357601002e-15
```

Although the function is for internal use only, fair enough about potential more general applicability. Even if it were only to be used for single precision, I haven't done any rigorous testing with precision. For what it's worth I think I honed in one the 1e-10 value when I set up `hypothesis` tests to compare results of the original transformations and the reimplemented ones.

But again, if there's fears of this leading to wrong results (fears that outweight the fears of computing Rodrigues' formula for no reason for identical cases) then we can just remove the fast path.

Reply:
If you remove the fast path, then a rotation angle of 0.0 does not give you an exact identity matrix. Despite this, I prefer this behavior over the built-in "fast path" because the output is smooth, and I can set up the "fast path" outside of the function call if need be. I could certainly be an exception.

For most applications I doubt that the "fast path" saves that much time because time is mostly spent in application of the transformation itself and not in building the transform matrix.

Reply:
Unlike the above examples, `sin(0) == 0` and `cos(0) == 1` are reliable truths. So no, without the fast path we still get an exact identity matrix for 0 angle. As I said, this is purely a performance optimization.

Partly why I went with this design is because _some people_ raised performance concerns early on :P And I just stuck with it. I think it makes no practical difference. I'll just change it to exact comparison with 0, which will still give us some true positives, without affecting near-zero values. Done in https://github.com/pyvista/pyvista/pull/1392/commits/eae56be3cb0b8bc5458b42c6de62378210bbad23.